aws/tagsCognito.go

```diff
@@ -0,0 +1,97 @@
+package aws
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cognitoidentity"
+	"github.com/hashicorp/terraform/helper/schema"
+	"log"
+)
+
+// setTags is a helper to set the tags for a resource. It expects the
+// tags field to be named "tags"
+func setTagsCognito(conn *cognitoidentity.CognitoIdentity, d *schema.ResourceData) error {
+	if d.HasChange("tags") {
+		oraw, nraw := d.GetChange("tags")
+		o := oraw.(map[string]interface{})
+		n := nraw.(map[string]interface{})
+		create, remove := diffTagsCognito(tagsFromMapCognito(o), tagsFromMapCognito(n))
+
+		// Set tags
+		if len(remove) > 0 {
+			log.Printf("[DEBUG] Removing tags: %s", remove)
+			tagsToRemove := &cognitoidentity.UntagResourceInput{
+				ResourceArn: aws.String(d.Get("arn").(string)),
+				TagKeys:     cognitoStringsToPointyString(remove),
+			}
+
+			_, err := conn.UntagResource(tagsToRemove)
+			if err != nil {
+				return err
+			}
+		}
+		if len(create) > 0 {
+			log.Printf("[DEBUG] Creating tags: %s", create)
+			tagsToAdd := &cognitoidentity.TagResourceInput{
+				ResourceArn: aws.String(d.Get("arn").(string)),
+				Tags:        cognitoTagsFromMap(create),
+			}
+			_, err := conn.TagResource(tagsToAdd)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func cognitoStringsToPointyString(s []string) []*string {
+	results := make([]*string, len(s))
+	for i, x := range s {
+		results[i] = aws.String(x)
+	}
+
+	return results
+}
+
+func cognitoTagsFromMap(responseTags map[string]string) map[string]*string {
+	results := make(map[string]*string, len(responseTags))
+	for k, v := range responseTags {
+		results[k] = aws.String(v)
+	}
+
+	return results
+}
+
+// diffTags takes our tags locally and the ones remotely and returns
+// the set of tags that must be created, and the set of tags that must
+// be destroyed.
+func diffTagsCognito(oldTags, newTags map[string]string) (map[string]string, []string) {
```

Comment:
This function should include unit testing to match the other `diffTags...` functions, see also: `aws/tagsDocDB_test.go`

Reply:
fixed in 4ed6457ff